File: OKXAutoSwapBot.py
import asyncio
import aiohttp
import json
import logging
import os
from datetime import datetime
from typing import Dict, Optional, Any
from web3 import Web3
from eth_account import Account
import time

class OKXAutoSwapBot:
    def __init__(self, config: Dict[str, Any]):
        self.config = config
        self.w3 = Web3(Web3.HTTPProvider(config['rpc_url']))
        self.account = Account.from_key(config['private_key'])
        self.wallet_address = self.account.address
        self.is_running = False
        self.logs = []
        
        # Token details
        self.token_address = '0x077574441c4f8763a37a2cfee2ecb444aa60a15e'
        self.chain_id = 42161  # Arbitrum One
        
        # API endpoints
        self.base_url = 'https://web3.okx.com/api/v5/dex'
        
        # Setup logging
        self.setup_logging()
        self.log('Bot initialized successfully')
    
    def setup_logging(self):
        """Setup logging configuration"""
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s - %(levelname)s - %(message)s',
            handlers=[
                logging.FileHandler('swap_logs.txt'),
                logging.StreamHandler()
            ]
        )
        self.logger = logging.getLogger(__name__)
    
    def log(self, message: str):
        """Log message with timestamp"""
        timestamp = datetime.now().isoformat()
        log_entry = f"[{timestamp}] {message}"
        print(log_entry)
        self.logs.append(log_entry)
        self.logger.info(message)
        
        # Save logs to file
        self.save_logs()
    
    def save_logs(self):
        """Save logs to file"""
        try:
            with open('swap_logs.txt', 'w') as f:
                f.write('\n'.join(self.logs))
        except Exception as e:
            print(f"Error saving logs: {e}")
    
    async def get_token_price(self, token_address: str) -> Optional[Dict[str, Any]]:
        """Get token price from OKX API"""
        try:
            async with aiohttp.ClientSession() as session:
                params = {
                    'chainId': self.chain_id,
                    'tokenContractAddress': token_address
                }
                
                async with session.get(f"{self.base_url}/aggregator/tokens", params=params) as response:
                    data = await response.json()
                    
                    if data.get('code') == '0' and data.get('data'):
                        return data['data']
                    
                    raise Exception('Failed to fetch token price')
                    
        except Exception as e:
            self.log(f"Error fetching token price: {str(e)}")
            return None
    
    async def get_swap_quote(self, from_token: str, to_token: str, amount: str) -> Optional[Dict[str, Any]]:
        """Get swap quote from OKX API"""
        try:
            async with aiohttp.ClientSession() as session:
                params = {
                    'chainId': self.chain_id,
                    'fromTokenAddress': from_token,
                    'toTokenAddress': to_token,
                    'amount': amount,
                    'slippage': self.config.get('slippage', 0.5),
                    'userWalletAddress': self.wallet_address
                }
                
                async with session.get(f"{self.base_url}/aggregator/quote", params=params) as response:
                    data = await response.json()
                    
                    if data.get('code') == '0' and data.get('data'):
                        return data['data']
                    
                    raise Exception('Failed to get swap quote')
                    
        except Exception as e:
            self.log(f"Error getting swap quote: {str(e)}")
            return None
    
    async def execute_swap(self, swap_data: Dict[str, Any]) -> Dict[str, Any]:
        """Execute swap transaction"""
        try:
            async with aiohttp.ClientSession() as session:
                payload = {
                    'chainId': self.chain_id,
                    'fromTokenAddress': swap_data['from_token'],
                    'toTokenAddress': swap_data['to_token'],
                    'amount': swap_data['amount'],
                    'slippage': self.config.get('slippage', 0.5),
                    'userWalletAddress': self.wallet_address,
                    'sort': 1,  # Sort by best price
                    'feePercent': 0
                }
                
                async with session.post(f"{self.base_url}/aggregator/swap", json=payload) as response:
                    data = await response.json()
                    
                    if data.get('code') == '0' and data.get('data'):
                        tx_data = data['data'][0]
                        
                        # Prepare transaction
                        transaction = {
                            'to': tx_data['to'],
                            'data': tx_data['data'],
                            'value': int(tx_data.get('value', '0')),
                            'gas': int(tx_data.get('gasLimit', '500000')),
                            'gasPrice': self.w3.to_wei('1', 'gwei'),
                            'nonce': self.w3.eth.get_transaction_count(self.wallet_address)
                        }
                        
                        # Sign and send transaction
                        signed_txn = self.account.sign_transaction(transaction)
                        tx_hash = self.w3.eth.send_raw_transaction(signed_txn.rawTransaction)
                        
                        self.log(f"Swap executed! TX Hash: {tx_hash.hex()}")
                        
                        # Wait for confirmation
                        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
                        self.log(f"Swap confirmed! Block: {receipt.blockNumber}")
                        
                        return {
                            'success': True,
                            'tx_hash': tx_hash.hex(),
                            'receipt': dict(receipt)
                        }
                    
                    raise Exception('Failed to execute swap')
                    
        except Exception as e:
            self.log(f"Error executing swap: {str(e)}")
            return {
                'success': False,
                'error': str(e)
            }
    
    async def check_price_conditions(self, current_price: float) -> Optional[Dict[str, str]]:
        """Check if price conditions are met"""
        conditions = self.config['conditions']
        
        # Check buy conditions
        if 'buy_when' in conditions:
            buy_conditions = conditions['buy_when']
            if 'price_below' in buy_conditions and current_price <= buy_conditions['price_below']:
                return {'action': 'buy', 'reason': f'Price below {buy_conditions["price_below"]}'}
        
        # Check sell conditions
        if 'sell_when' in conditions:
            sell_conditions = conditions['sell_when']
            if 'price_above' in sell_conditions and current_price >= sell_conditions['price_above']:
                return {'action': 'sell', 'reason': f'Price above {sell_conditions["price_above"]}'}
        
        return None
    
    async def get_wallet_balance(self, token_address: str) -> str:
        """Get wallet balance for ETH or token"""
        try:
            if token_address == 'ETH' or token_address == '0x0000000000000000000000000000000000000000':
                balance = self.w3.eth.get_balance(self.wallet_address)
                return str(self.w3.from_wei(balance, 'ether'))
            else:
                # Token balance using contract call
                contract_abi = [
                    {
                        "constant": True,
                        "inputs": [{"name": "_owner", "type": "address"}],
                        "name": "balanceOf",
                        "outputs": [{"name": "balance", "type": "uint256"}],
                        "type": "function"
                    },
                    {
                        "constant": True,
                        "inputs": [],
                        "name": "decimals",
                        "outputs": [{"name": "", "type": "uint8"}],
                        "type": "function"
                    }
                ]
                
                contract = self.w3.eth.contract(address=token_address, abi=contract_abi)
                balance = contract.functions.balanceOf(self.wallet_address).call()
                decimals = contract.functions.decimals().call()
                
                return str(balance / (10 ** decimals))
                
        except Exception as e:
            self.log(f"Error getting wallet balance: {str(e)}")
            return '0'
    
    async def monitor_and_swap(self):
        """Main monitoring loop"""
        self.log('Starting price monitoring...')
        
        while self.is_running:
            try:
                # Get current token price
                token_data = await self.get_token_price(self.token_address)
                
                if not token_data:
                    await asyncio.sleep(self.config.get('check_interval', 30))
                    continue
                
                current_price = float(token_data['price'])
                self.log(f"Current token price: ${current_price}")
                
                # Check if conditions are met
                condition = await self.check_price_conditions(current_price)
                
                if condition:
                    self.log(f"Condition met: {condition['reason']} - Action: {condition['action']}")
                    
                    # Get wallet balances
                    eth_balance = await self.get_wallet_balance('ETH')
                    token_balance = await self.get_wallet_balance(self.token_address)
                    
                    self.log(f"ETH Balance: {eth_balance}, Token Balance: {token_balance}")
                    
                    # Execute swap based on condition
                    if condition['action'] == 'buy':
                        swap_amount = str(int(self.w3.to_wei(self.config['swap_amount'], 'ether')))
                        
                        quote = await self.get_swap_quote(
                            '0x0000000000000000000000000000000000000000',  # ETH
                            self.token_address,
                            swap_amount
                        )
                        
                        if quote:
                            result = await self.execute_swap({
                                'from_token': '0x0000000000000000000000000000000000000000',
                                'to_token': self.token_address,
                                'amount': swap_amount
                            })
                            
                            if result['success']:
                                self.log('Buy order executed successfully!')
                    
                    elif condition['action'] == 'sell':
                        token_balance_wei = int(float(token_balance) * (10 ** 18))
                        swap_amount = str(int(token_balance_wei * self.config.get('sell_percentage', 100) / 100))
                        
                        quote = await self.get_swap_quote(
                            self.token_address,
                            '0x0000000000000000000000000000000000000000',  # ETH
                            swap_amount
                        )
                        
                        if quote:
                            result = await self.execute_swap({
                                'from_token': self.token_address,
                                'to_token': '0x0000000000000000000000000000000000000000',
                                'amount': swap_amount
                            })
                            
                            if result['success']:
                                self.log('Sell order executed successfully!')
                
                # Wait before next check
                await asyncio.sleep(self.config.get('check_interval', 30))
                
            except Exception as e:
                self.log(f"Error in monitoring loop: {str(e)}")
                await asyncio.sleep(5)
    
    def start(self):
        """Start the bot"""
        if self.is_running:
            self.log('Bot is already running')
            return
        
        self.is_running = True
        self.log('Starting OKX Auto Swap Bot...')
        asyncio.run(self.monitor_and_swap())
    
    def stop(self):
        """Stop the bot"""
        if not self.is_running:
            self.log('Bot is not running')
            return
        
        self.is_running = False
        self.log('Stopping OKX Auto Swap Bot...')
    
    def get_status(self) -> Dict[str, Any]:
        """Get bot status"""
        return {
            'is_running': self.is_running,
            'wallet_address': self.wallet_address,
            'token_address': self.token_address,
            'chain_id': self.chain_id,
            'logs_count': len(self.logs)
        }

if __name__ == "__main__":
    # This will be imported from config.py in production
    pass

File: OKXAutoSwapBot.js
const axios = require('axios');
const { ethers } = require('ethers');
const fs = require('fs');
const path = require('path');

class OKXAutoSwapBot {
    constructor(config) {
        this.config = config;
        this.provider = new ethers.providers.JsonRpcProvider(config.rpcUrl);
        this.wallet = new ethers.Wallet(config.privateKey, this.provider);
        this.isRunning = false;
        this.logs = [];
        
        // Token details
        this.tokenAddress = '0x077574441c4f8763a37a2cfee2ecb444aa60a15e';
        this.chainId = 42161; // Arbitrum One
        
        // API endpoints
        this.baseURL = 'https://web3.okx.com/api/v5/dex';
        
        this.log('Bot initialized successfully');
    }

    log(message) {
        const timestamp = new Date().toISOString();
        const logEntry = `[${timestamp}] ${message}`;
        console.log(logEntry);
        this.logs.push(logEntry);
        
        // Save logs to file
        this.saveLogs();
    }

    saveLogs() {
        const logsPath = path.join(__dirname, 'swap_logs.txt');
        fs.writeFileSync(logsPath, this.logs.join('\n'));
    }

    async getTokenPrice(tokenAddress) {
        try {
            const response = await axios.get(`${this.baseURL}/aggregator/tokens`, {
                params: {
                    chainId: this.chainId,
                    tokenContractAddress: tokenAddress
                }
            });
            
            if (response.data.code === '0' && response.data.data) {
                return response.data.data;
            }
            
            throw new Error('Failed to fetch token price');
        } catch (error) {
            this.log(`Error fetching token price: ${error.message}`);
            return null;
        }
    }

    async getSwapQuote(fromToken, toToken, amount) {
        try {
            const response = await axios.get(`${this.baseURL}/aggregator/quote`, {
                params: {
                    chainId: this.chainId,
                    fromTokenAddress: fromToken,
                    toTokenAddress: toToken,
                    amount: amount,
                    slippage: this.config.slippage || 0.5,
                    userWalletAddress: this.wallet.address
                }
            });
            
            if (response.data.code === '0' && response.data.data) {
                return response.data.data;
            }
            
            throw new Error('Failed to get swap quote');
        } catch (error) {
            this.log(`Error getting swap quote: ${error.message}`);
            return null;
        }
    }

    async executeSwap(swapData) {
        try {
            const response = await axios.post(`${this.baseURL}/aggregator/swap`, {
                chainId: this.chainId,
                fromTokenAddress: swapData.fromToken,
                toTokenAddress: swapData.toToken,
                amount: swapData.amount,
                slippage: this.config.slippage || 0.5,
                userWalletAddress: this.wallet.address,
                sort: 1, // Sort by best price
                feePercent: 0
            });
            
            if (response.data.code === '0' && response.data.data) {
                const txData = response.data.data[0];
                
                // Execute the transaction
                const tx = {
                    to: txData.to,
                    data: txData.data,
                    value: txData.value || '0',
                    gasLimit: txData.gasLimit || '500000',
                    gasPrice: ethers.utils.parseUnits('1', 'gwei')
                };
                
                const transaction = await this.wallet.sendTransaction(tx);
                this.log(`Swap executed! TX Hash: ${transaction.hash}`);
                
                // Wait for confirmation
                const receipt = await transaction.wait();
                this.log(`Swap confirmed! Block: ${receipt.blockNumber}`);
                
                return {
                    success: true,
                    txHash: transaction.hash,
                    receipt: receipt
                };
            }
            
            throw new Error('Failed to execute swap');
        } catch (error) {
            this.log(`Error executing swap: ${error.message}`);
            return {
                success: false,
                error: error.message
            };
        }
    }

    async checkPriceConditions(currentPrice) {
        const conditions = this.config.conditions;
        
        // Check buy conditions
        if (conditions.buyWhen) {
            if (conditions.buyWhen.priceBelow && currentPrice <= conditions.buyWhen.priceBelow) {
                return { action: 'buy', reason: `Price below ${conditions.buyWhen.priceBelow}` };
            }
            
            if (conditions.buyWhen.priceIncrease) {
                // Check for price increase percentage
                const increasePercent = conditions.buyWhen.priceIncrease;
                // This would need historical price tracking
            }
        }
        
        // Check sell conditions
        if (conditions.sellWhen) {
            if (conditions.sellWhen.priceAbove && currentPrice >= conditions.sellWhen.priceAbove) {
                return { action: 'sell', reason: `Price above ${conditions.sellWhen.priceAbove}` };
            }
            
            if (conditions.sellWhen.priceDecrease) {
                // Check for price decrease percentage
                const decreasePercent = conditions.sellWhen.priceDecrease;
                // This would need historical price tracking
            }
        }
        
        return null;
    }

    async getWalletBalance(tokenAddress) {
        try {
            if (tokenAddress === 'ETH' || tokenAddress === ethers.constants.AddressZero) {
                const balance = await this.wallet.getBalance();
                return ethers.utils.formatEther(balance);
            } else {
                const tokenContract = new ethers.Contract(
                    tokenAddress,
                    ['function balanceOf(address) view returns (uint256)', 'function decimals() view returns (uint8)'],
                    this.provider
                );
                
                const balance = await tokenContract.balanceOf(this.wallet.address);
                const decimals = await tokenContract.decimals();
                
                return ethers.utils.formatUnits(balance, decimals);
            }
        } catch (error) {
            this.log(`Error getting wallet balance: ${error.message}`);
            return '0';
        }
    }

    async monitorAndSwap() {
        this.log('Starting price monitoring...');
        
        while (this.isRunning) {
            try {
                // Get current token price
                const tokenData = await this.getTokenPrice(this.tokenAddress);
                
                if (!tokenData) {
                    await this.sleep(this.config.checkInterval || 30000);
                    continue;
                }
                
                const currentPrice = parseFloat(tokenData.price);
                this.log(`Current token price: $${currentPrice}`);
                
                // Check if conditions are met
                const condition = await this.checkPriceConditions(currentPrice);
                
                if (condition) {
                    this.log(`Condition met: ${condition.reason} - Action: ${condition.action}`);
                    
                    // Get wallet balances
                    const ethBalance = await this.getWalletBalance('ETH');
                    const tokenBalance = await this.getWalletBalance(this.tokenAddress);
                    
                    this.log(`ETH Balance: ${ethBalance}, Token Balance: ${tokenBalance}`);
                    
                    // Execute swap based on condition
                    if (condition.action === 'buy') {
                        const swapAmount = ethers.utils.parseEther(this.config.swapAmount.toString());
                        
                        const quote = await this.getSwapQuote(
                            ethers.constants.AddressZero, // ETH
                            this.tokenAddress,
                            swapAmount.toString()
                        );
                        
                        if (quote) {
                            const result = await this.executeSwap({
                                fromToken: ethers.constants.AddressZero,
                                toToken: this.tokenAddress,
                                amount: swapAmount.toString()
                            });
                            
                            if (result.success) {
                                this.log(`Buy order executed successfully!`);
                            }
                        }
                    } else if (condition.action === 'sell') {
                        const tokenBalanceWei = ethers.utils.parseUnits(tokenBalance, 18);
                        const swapAmount = tokenBalanceWei.mul(this.config.sellPercentage || 100).div(100);
                        
                        const quote = await this.getSwapQuote(
                            this.tokenAddress,
                            ethers.constants.AddressZero, // ETH
                            swapAmount.toString()
                        );
                        
                        if (quote) {
                            const result = await this.executeSwap({
                                fromToken: this.tokenAddress,
                                toToken: ethers.constants.AddressZero,
                                amount: swapAmount.toString()
                            });
                            
                            if (result.success) {
                                this.log(`Sell order executed successfully!`);
                            }
                        }
                    }
                }
                
                // Wait before next check
                await this.sleep(this.config.checkInterval || 30000);
                
            } catch (error) {
                this.log(`Error in monitoring loop: ${error.message}`);
                await this.sleep(5000);
            }
        }
    }

    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    start() {
        if (this.isRunning) {
            this.log('Bot is already running');
            return;
        }
        
        this.isRunning = true;
        this.log('Starting OKX Auto Swap Bot...');
        this.monitorAndSwap();
    }

    stop() {
        if (!this.isRunning) {
            this.log('Bot is not running');
            return;
        }
        
        this.isRunning = false;
        this.log('Stopping OKX Auto Swap Bot...');
    }

    getStatus() {
        return {
            isRunning: this.isRunning,
            walletAddress: this.wallet.address,
            tokenAddress: this.tokenAddress,
            chainId: this.chainId,
            logsCount: this.logs.length
        };
    }
}

// Configuration example
const config = {
    // Required: Your private key (KEEP THIS SECURE!)
    privateKey: 'YOUR_PRIVATE_KEY_HERE',
    
    // Required: Arbitrum One RPC URL
    rpcUrl: 'https://arb1.arbitrum.io/rpc',
    
    // Swap conditions
    conditions: {
        buyWhen: {
            priceBelow: 0.001, // Buy when price drops below this value
            // priceIncrease: 5 // Buy when price increases by this percentage
        },
        sellWhen: {
            priceAbove: 0.01, // Sell when price goes above this value
            // priceDecrease: 10 // Sell when price decreases by this percentage
        }
    },
    
    // Swap settings
    swapAmount: 0.01, // Amount of ETH to swap when buying
    sellPercentage: 100, // Percentage of tokens to sell (100 = all tokens)
    slippage: 0.5, // Slippage tolerance in percentage
    
    // Monitoring settings
    checkInterval: 30000, // Check price every 30 seconds
};

// Usage example
const bot = new OKXAutoSwapBot(config);

// Event handlers
process.on('SIGINT', () => {
    console.log('\nReceived SIGINT, stopping bot...');
    bot.stop();
    process.exit(0);
});

process.on('SIGTERM', () => {
    console.log('\nReceived SIGTERM, stopping bot...');
    bot.stop();
    process.exit(0);
});

// Export for use as module
module.exports = OKXAutoSwapBot;

// If running directly
if (require.main === module) {
    console.log('=== OKX Auto Swap Bot for Arbitrum One ===');
    console.log('Token Address:', config.tokenAddress);
    console.log('Wallet Address:', bot.wallet.address);
    console.log('=========================================');
    
    // Start the bot
    bot.start();
}
